### PR TITLE
Add benchmark

### PR DESCRIPTION
**Pull request by benraskin92**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,6 +52,33 @@ Run Prometheus Docker image:
 
     $ docker run -p 9090:9090 -v $GOPATH/src/github.com/m3db/m3coordinator/docker/prometheus-mac.yml:/etc/prometheus/prometheus.yml quay.io/prometheus/prometheus
 
+### Benchmarking
```
Comment (nikunjgit) [inline README.md]: instead of putting this here, you can put a README.md file in benchmarks directory

File: benchmark/data/data_gen.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#!/bin/sh
```
Comment (nikunjgit) [inline benchmark/data/data_gen.sh]: some comments ?

File: storage/convertor.go [hunk 3]
```diff
@@ -155,3 +155,12 @@ func SeriesToPromSamples(series *ts.Series) []*prompb.Sample {
 	}
 	return samples
 }
+
+// FromMapToTags converts a map[string]string to Tags
+func FromMapToTags(mapTags map[string]string) models.Tags {
```
Comment (nikunjgit) [inline storage/convertor.go]: remove ?

File: benchmark/read/id.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+// Metrics is the OpenTSDB style metrics
+type Metrics struct {
+	Name string            `json:"metric"`
+	Tags map[string]string `json:"tags"`
+}
+
+var sortedKeys []string
+var buffer = bytes.NewBuffer(nil)
+
+func getIDs(fileName string) []string {
```
Comment (nikunjgit) [inline benchmark/read/id.go]: use the parallel implementation from write ?

File: benchmark/read/query_benchmark.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	timestampStart = timestampStart.UTC()
+	timestampEnd, err = time.Parse(time.RFC3339, timestampEndStr)
+	if err != nil {
+		log.Fatal(err)
+	}
+	timestampEnd = timestampEnd.UTC()
+
+	m3dbClientOpts := cfg.M3DBClientCfg
+	m3dbClient, err := m3dbClientOpts.NewClient(client.ConfigurationParameters{}, func(v client.Options) client.Options {
+		return v.SetMinConnectionCount(1).SetMaxConnectionCount(1).SetWriteBatchSize(batch).SetWriteOpPoolSize(batch * 2)
```
Comment (nikunjgit) [inline benchmark/read/query_benchmark.go]: use the settings from write ?

File: benchmark/write/convert_to_m3.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	fd, err := os.OpenFile(fileName, os.O_RDONLY, 0)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Unable to read json file, got error: %v", err)
+		os.Exit(1)
+	}
+
+	defer fd.Close()
+
+	var (
+		metrics = make([]*M3Metric, 0, 100000)
```
Comment (nikunjgit) [inline benchmark/write/convert_to_m3.go]: should size of all these be initialized to workers ? If not, then maybe move this number as a constant outside and use everywhere

File: benchmark/write/convert_to_m3.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+		scanner = bufio.NewScanner(fd)
+	)
+	var wg sync.WaitGroup
+	dataChannel := make(chan []byte, 100000)
+	metricChannel := make(chan *M3Metric, 100000)
+	for w := 0; w < workers; w++ {
+		wg.Add(1)
+		go func() {
```
Comment (nikunjgit) [inline benchmark/write/convert_to_m3.go]: move this out into its own func ?

File: benchmark/write/write_benchmark.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+		}()
+	}
+
+	fmt.Printf("waiting for workers to spin up...\n")
+	waitForInit.Wait()
+	fmt.Printf("done\n")
+
+	b := &benchmarker{address: address, benchmarkers: benchmarkers}
+	go b.serve()
+	go b.queryBenchmarkers()
```
Comment (nikunjgit) [inline benchmark/write/write_benchmark.go]: probably remove this line until we fix the bug for wrong results

File: benchmark/write/write_benchmark.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+	<-inputDone
+	close(ch)
+
+	wg.Wait()
+	sum := 0
+	for i := 0; i < workers; i++ {
+		sum += <-itemsWritten
+	}
+
+	fmt.Println(sum)
```
Comment (nikunjgit) [inline benchmark/write/write_benchmark.go]: remove ?

File: benchmark/write/write_benchmark.go [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+	var itemsWritten int
+	// var otherWG sync.WaitGroup
+	for query := range ch {
+		id := query.ID
+		if err := session.Write(namespace, id, query.Time, query.Value, xtime.Millisecond, nil); err != nil {
+			fmt.Println(err)
+		} else {
+			stat.incWrites()
+		}
+		// otherWG.Add(1)
```
Comment (nikunjgit) [inline benchmark/write/write_benchmark.go]: remove ?

File: benchmark/README.MD [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+3) Create metrics - from the `github.com/m3db/m3coordinator/benchmark/data` directory, run:
+
+       $ ./data_gen.sh
+
+> Note: If you need to adjust the script to get more metrics, make sure the timestamp start and end are within the `buffer_past` config that is set for m3db and the current time, otherwise you will get datapoint too far in the past/future errors.
+
+4) Start m3db. You can use the config that is in `github.com/m3db/m3coordinator/benchmark/`
+
+        $ ./bin/m3dbnode -f ~/gocode/src/github.com/m3db/m3coordinator/benchmark/m3db_config.yaml
```
Comment (nikunjgit) [inline benchmark/README.MD]: change this to $GOPATH/src/github.com/m3db/m3coordinator/benchmark/m3db_config.yaml

File: benchmark/README.MD [hunk 12]
```diff
@@ -0,0 +1,26 @@
+### Benchmarking
+
+To benchmark m3db writes using m3coordinator.
+
+1) Make sure you have the lastest version of m3coordinator
+2) Pull down benraskin92's fork of `influxdb-comparisons`:
+
+       $ git clone https://github.com/benraskin92/influxdb-comparisons.git
```
Comment (nikunjgit) [inline benchmark/README.MD]: probably need to clarify where to clone

File: benchmark/common/parse_json.go [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+func ConvertToM3(fileName string, workers int, f func(*M3Metric)) {
+	fd, err := os.OpenFile(fileName, os.O_RDONLY, 0)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Unable to read json file, got error: %v", err)
+		os.Exit(1)
+	}
+
+	defer fd.Close()
+
+	var (
```
Comment (nikunjgit) [inline benchmark/common/parse_json.go]: just use assignment ?

File: benchmark/common/parse_json.go [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+	for scanner.Scan() {
+		data := bytes.TrimSpace(scanner.Bytes())
+		b := make([]byte, len(data))
+		copy(b, data)
+		dataChannel <- b
+	}
+
+	close(dataChannel)
+
```
Comment (nikunjgit) [inline benchmark/common/parse_json.go]: remove spaces ?

File: benchmark/common/parse_json.go [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+	}
+}
+
+func unmarshalMetrics(dataChannel chan []byte, metricChannel chan *M3Metric) {
+	for data := range dataChannel {
+		if len(data) != 0 {
+			var m Metrics
+			if err := json.Unmarshal(data, &m); err != nil {
+				fmt.Fprintf(os.Stderr, "Unable to unmarshal json, got error: %v", err)
+				os.Exit(1)
```
Comment (nikunjgit) [inline benchmark/common/parse_json.go]: use panic ?

File: benchmark/README.MD [hunk 16]
```diff
@@ -0,0 +1,26 @@
+### Benchmarking
+
+To benchmark m3db writes using m3coordinator.
```
Comment (arnikola) [inline benchmark/README.MD]: Can this just be consolidated into a bash script file?

File: benchmark/common/parse_json.go [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	defer fd.Close()
+
+	scanner := bufio.NewScanner(fd)
+
+	dataChannel := make(chan []byte, MetricsLen)
+	metricChannel := make(chan *M3Metric, MetricsLen)
+	for w := 0; w < workers; w++ {
+		wg.Add(1)
```
Comment (arnikola) [inline benchmark/common/parse_json.go]: can we do something like

```go
wg.Add(1)
go func(dc chan []byte, mc chan *M3Metric) {
	defer wg.Done()
	unmarshalMetrics(dc, mc)
}(dataChannel, metricChannel)
```

then make `wg` scoped to the function?

File: benchmark/common/parse_json.go [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+	close(dataChannel)
+	wg.Wait()
+	close(metricChannel)
+	if err := scanner.Err(); err != nil {
+		panic(err)
+	}
+}
+
+func unmarshalMetrics(dataChannel chan []byte, metricChannel chan *M3Metric) {
```
Comment (arnikola) [inline benchmark/common/parse_json.go]: `dataChannel chan<- []byte, metricChannel<- chan *M3Metric`

Not sure if we really bother with this in real life or if it's just something that go tutorials like to point out 😜

File: benchmark/common/parse_json.go [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+			}
+
+			metricChannel <- &M3Metric{ID: id(m.Tags, m.Name), Time: storage.PromTimestampToTime(m.Time), Value: m.Value}
+		}
+	}
+	wg.Done()
+}
+
+func id(lowerCaseTags map[string]string, name string) string {
+	var sortedKeys []string
```
Comment (arnikola) [inline benchmark/common/parse_json.go]: `keys := make([]int, len(lowerCaseTags))` for no append resizing?

File: benchmark/common/parse_json.go [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+	buffer.WriteString(strings.ToLower(name))
+
+	// Generate tags in alphabetical order & write to buffer
+	i := 0
+	for key := range lowerCaseTags {
+		sortedKeys = append(sortedKeys, key)
+		i++
+	}
+	sort.Strings(sortedKeys)
+
```
Comment (arnikola) [inline benchmark/common/parse_json.go]: `for _, key := range sortedKeys {`

Is it better to write both strings in a single buffer.WriteString?

File: benchmark/read/query_benchmark.go [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+	flag.StringVar(&timestampEndStr, "timestamp-end", "2016-01-01T06:00:00Z", "Ending timestamp (RFC3339).")
+	flag.BoolVar(&memprofile, "memprofile", false, "Enable memory profile")
+	flag.BoolVar(&cpuprofile, "cpuprofile", false, "Enable cpu profile")
+	flag.Parse()
+}
+
+func main() {
+	var cfg config.Configuration
+	if err := xconfig.LoadFile(&cfg, m3dbClientCfg); err != nil {
+		fmt.Fprintf(os.Stderr, "Unable to load %s: %v", m3dbClientCfg, err)
```
Comment (arnikola) [inline benchmark/read/query_benchmark.go]: I think these can be `log.Fatalf(..`

File: benchmark/write/benchmarker.go [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+
+func (s *stats) setRunTimeMs(v int64) {
+	atomic.StoreInt64(&s.RunTimeMs, v)
+}
+
+func (s *stats) snapshot() stats {
+	return stats{Writes: s.getWrites(), RunTimeMs: s.getRunTimeMs()}
+}
+
+// Do we need this? It's not used.
```
Comment (arnikola) [inline benchmark/write/benchmarker.go]: delete?

File: benchmark/write/benchmarker.go [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+	if err := http.ListenAndServe(b.address, mux); err != nil {
+		fmt.Fprintf(os.Stderr, "server could not listen on %s: %v", b.address, err)
+	}
+}
+
+func (b *benchmarker) allAddresses() []string {
+	var all []string
+	for _, addr := range strings.Split(b.benchmarkers, ",") {
+		addr = strings.TrimSpace(addr)
+		if addr == "" {
```
Comment (arnikola) [inline benchmark/write/benchmarker.go]: if addr != "" {
  all = append(all,addr)
}

File: benchmark/write/write_benchmark.go [hunk 24]
```diff
@@ -0,0 +1,175 @@
...
+	metrics := make([]*common.M3Metric, 0, common.MetricsLen)
+	common.ConvertToM3(dataFile, workers, func(m *common.M3Metric) {
+		metrics = append(metrics, m)
+	})
+	ch := make(chan *common.M3Metric, workers)
+	inputDone = make(chan struct{})
+
+	var cfg config.Configuration
+	if err := xconfig.LoadFile(&cfg, m3dbClientCfg); err != nil {
+		fmt.Fprintf(os.Stderr, "Unable to load %s: %v", m3dbClientCfg, err)
```
Comment (arnikola) [inline benchmark/write/write_benchmark.go]: log.Fatalf here too

**Pull request merged**